docs/specs/consensus/README.md

```diff
@@ -0,0 +1,1111 @@
+
+The relative minimum window density of blocks `B1` and `B2` is defined as.
+
+```rust
+fn relativeMinWindowDensity(B1, B2) -> u32
+{
+    let max_slot = max(globalSlot(B1), globalSlot(B2))
+
+    // Grace-period rule
+    if max_slot < grace_period_end {
+        return B1.min_window_density
+    }
+
+    // Compute B1's window projected to max_slot
+    let projected_window = {
+        // Compute shift count
+        let shift_count = min(max(max_slot - B1.curr_global_slot - 1, 0), sub_windows_per_window)
+
+        // Initialize projected window
+        let projected_window = B1.sub_window_densities
+
+        // Ring-shift
+        let i = relativeSubWindow(B1.curr_global_slot)
+        while shift_count > 0 {
+          i = i + 1 mod sub_windows_per_window
+          projected_window[i] = 0
+          shift_count--;
+        }
+
+        return projected_window
+    }
+
+    // Compute projected window density
+    let projected_window_density = density(projected_window)
+
+    // Compute minimum window density
+    return min(B1.min_window_density, projected_window_density)
+}
+```
+
+This description was adopted to aid understanding, providing explanations where possible of why strategies have been adopted and what conditions are important.  In a production implementation consideration must be given to performance implications and another implementation may be desirable.
+
+# 6 Protocol
+
+This section specifies the consensus protocol in terms of events and how they MUST be implemented by a compatible peer.  The required events are:
+
+* [`Initialize consensus`](#61-initialize-consensus)
+* [`Select chain`](#62-select-chain)
+
+Additionally there are certain local data members that all peers MUST maintain in order to participate in consensus.
+
+| Parameter   | Description |
+| - | - |
+| `genesis_block` | The initial block in the blockchain |
+| `neighbors`     | Set of connections to neighboring peers |
+| `chains`        | Set of known (succinct) candidate chains |
+| `tip`           | Currently selected chain according to the chain selection algorithm (i.e. secure chain) |
+
+How these are represented is up to the implementation, but careful consideration must be given to scalability.
+
+In the following description we use _dot notation_ to refer the local data members of peers. For example, given peer `P`, we use `P.genesis_block` and `P.tip`, to refer to the genesis block and currently selected chain, respectively.
+
+## 6.1 Initialize consensus
+
+Things a peer MUST do to initialize consensus includes
+
+* Load the genesis block
+* Get head of the current chain
```

Comment:
This involves a proof of finality where we get the head of a chain and the `max(k,B.height-1)`th block back from that head.

Reply:
It sounds like there is an algorithm needed here.

Does this belong in consensus or could it be something consensus calls for from another layer (e.g. the p2p layer)?

Reply:
I think this should be part of another section, perhaps related to P2P algorithms.  Also, for the 3rd party rust implementation we are now considering a non-p2p approach initially, involving relay peers.  So it would be good if this "layer" were pluggable for different approaches, as now it seems we have at least two.

Reply:
I think this belong in consensus, because this proof is necessary to connect the head of the current chain back to a finalized block. This is a detail that should not change in alternative implementations.

Reply:
I think you mean `min(k,B.height-1)`, right?